pkg/await/apps_statefulset_test.go

```diff
@@ -0,0 +1,1252 @@
+	}{
+		{
+			description: "Read should fail if StatefulSet status empty",
+			statefulset: statefulsetAdded,
+			expectedSubErrors: []string{
+				"Failed to observe the expected number of ready replicas",
+				".status.currentRevision does not match .status.updateRevision",
+			},
+		},
+		{
+			description: "Read should fail if StatefulSet is progressing",
+			statefulset: statefulsetProgressing,
+			expectedSubErrors: []string{
+				"Failed to observe the expected number of ready replicas",
+			},
+		},
+		{
+			description: "Read should succeed if StatefulSet is ready",
+			statefulset: statefulsetReady,
+		},
+		{
+			description: "Read should succeed if StatefulSet update is ready",
+			statefulset: statefulsetUpdateSuccess,
+		},
+	}
+
+	for _, test := range tests {
+		awaiter := makeStatefulSetInitAwaiter(
+			updateAwaitConfig{
+				createAwaitConfig: mockAwaitConfig(statefulsetInput(inputNamespace, inputName, targetService)),
+			})
+		statefulset := test.statefulset(inputNamespace, inputName, targetService)
+		err := awaiter.read(statefulset, unstructuredList())
+		if test.expectedSubErrors != nil {
+			assert.Equal(t, test.expectedSubErrors, err.(*initializationError).SubErrors(), test.description)
+		} else {
+			assert.Nil(t, err, test.description)
+		}
+	}
+}
+
+// --------------------------------------------------------------------------
+
+// StatefulSet objects.
+
+// --------------------------------------------------------------------------
+
+func statefulsetInput(namespace, name, targetService string) *unstructured.Unstructured {
+	obj, err := decodeUnstructured(fmt.Sprintf(`{
+    "kind": "StatefulSet",
+    "apiVersion": "apps/v1beta1",
+    "metadata": {
+        "namespace": "%s",
+        "name": "%s",
+        "labels": {
+            "app": "foo"
+        }
+    },
+    "spec": {
+        "replicas": 2,
+        "selector": {
+            "matchLabels": {
+                "app": "foo"
+            }
+        },
+		"serviceName": "%s",
+        "template": {
+            "metadata": {
+                "labels": {
+                    "app": "foo"
+                }
+            },
+            "spec": {
+                "containers": [
+                    {
+                        "name": "nginx",
+                        "image": "nginx",
+						"volumeMounts": [
+							{
+								"mountPath": "/usr/share/nginx/html",
+								"name": "www"
+							}
+						]
+                    }
+                ],
+				"terminationGracePeriodSeconds": 10
+            }
+        },
+		"volumeClaimTemplates": [
+			{
+				"metadata": {
+					"name": "www"
+				},
+				"spec": {
+					"accessModes": [
+						"ReadWriteOnce"
+					],
+					"resources": {
+						"requests": {
+							"storage": "1Gi"
+						}
+					}
+				}
+			}
+		]
+    }
+}`, namespace, name, targetService))
+	if err != nil {
+		panic(err)
+	}
+	return obj
+}
+
+func statefulsetAdded(namespace, name, targetService string) *unstructured.Unstructured {
```

Comment:
The difference between each of these is kind of subtle -- what do you think about putting a little comment above each to point out the major differences?